### PR TITLE
Ansible 2.10

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

- Updates ansible to 2.10 (`python3Package.ansible-base`)
- Updates ansible_2_9 to 2.9.21 (`python3Packages.ansible` for `ansible<3.0`)
- Drops ansible_2_8 (EOL)
- Packaged `ansible==3.4.0` (collections) as `ansible-collections`, because we still rely on `python3Packages.ansible` for `ansible_2_9`, the upstream made a mess.
- Packaged all missing dependencies for collections maintained by the upstream (`ansible.{netcommon,posix,utils,windows}`), as well as those for junos and netbox collections, which I use personally for work
- And one more fun fact: Ansible 2.11 will be `ansible-core` :facepalm: 

Supersedes #118290, #100147.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/jxmlease/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  ];
+
+  checkPhase = ''
+    runHook preCheck
+    ${python.interpreter} -m unittest discover -v
+    runHook postCheck
+  '';
+
+  meta = with lib; {
+    description = "jxmlease converts between XML and intelligent Python data structures";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jxmlease/default.nix:28]:
```suggestion
    description = "Converts between XML and intelligent Python data structures";
```

File: pkgs/development/python-modules/pynetbox/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+buildPythonPackage rec {
+  pname = "pynetbox";
+  version = "6.1.2";
+
+  src = fetchFromGitHub {
+    owner = "netbox-community";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "0di07rny3gqdfb0rf7hm3x03rpn7rydpv3lrl7cak2ccpqm0wzhl";
+    leaveDotGit = true;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pynetbox/default.nix:19]: Can we set SETUPTOOLS_SCM_PRETEND_VERSION instead?

File: pkgs/tools/admin/ansible/default.nix [hunk 3]
```diff
@@ -3,30 +3,8 @@
...
-      rev = "v${version}";
-      sha256 = "0k9rs5ajx0chaq0xr1cj4x7fr5n8kd4y856miss6k01iv2m7yx42";
-    };
-  }));
+  ansible_2_10 = python3Packages.toPythonApplication python3Packages.ansible-base;
 
+  # End of support 2021/10/02, End of life 2021/12/31
   ansible_2_9 = python3Packages.toPythonApplication python3Packages.ansible;
-
-  ansible_2_8 = python3Packages.toPythonApplication (python3Packages.ansible.overridePythonAttrs (old: rec {
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/ansible/default.nix:23]: We should add an alias for this.

File: pkgs/development/python-modules/ttp/templates.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+}:
+
+buildPythonPackage rec {
+  pname = "ttp_templates";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/ttp/templates.nix:7]:
```suggestion
  pname = "ttp-templates";
```

File: pkgs/development/python-modules/ttp/templates.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  version = "0.1.1";
+  format = "setuptools";
+
+  src = fetchPypi {
+    inherit pname version;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/ttp/templates.nix:12]:
```suggestion
    pname = "ttp_templates";
    inherit version;
```

File: pkgs/development/python-modules/pynetbox/default.nix (hunk 2 above)
Comment (mweinelt) [reply pkgs/development/python-modules/pynetbox/default.nix:19]: Neat! Thanks for the tip.

File: pkgs/tools/admin/ansible/default.nix (hunk 3 above)
Comment (mweinelt) [reply pkgs/tools/admin/ansible/default.nix:23]: I'll add a throw instead of removing it, and will remove the throw after branch-off. Nobody should be surprised that Ansible 2.8 went EOL tbh.

File: pkgs/development/python-modules/ansible/base.nix [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      --replace "[python," "["
+  '';
+
+  nativeBuildInputs = [
+    installShellFiles
+  ];
+
+  propagatedBuildInputs = [
+    # depend on ansible-collections instead of the other way around
+    ansible-collections
```
Comment (mweinelt) [inline pkgs/development/python-modules/ansible/base.nix:46]: Hm, the current way I structured this is that overrides need to be applied onto `ansible-collections`, to get the required packages for the relevant collections. I quickly checked, and ansible-collections has a bit over 100 dependencies, some of which we don't have packaged, so I went this route.

Now I wonder whether overriding is easy to accomplish for mere mortals.

File: pkgs/development/python-modules/ansible/collections.nix [hunk 7]
```diff
@@ -0,0 +1,72 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, ansible-base
+# sorted by collection
+, withJunos ? false
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/ansible/collections.nix:6]:
```suggestion
# sorted by collection
, jsonschema
, withJunos ? false
```

File: pkgs/development/python-modules/ansible/collections.nix [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    jxmlease
+    ncclient
+    paramiko
+    scp
+    xmltodict
+  ] ++ lib.optionals (withNetbox) [
+    pynetbox
+  ]);
+
+  dontStrip = true;
```
Comment (jonringer) [inline pkgs/development/python-modules/ansible/collections.nix:64]: why is this needed?

**Review by jonringer — APPROVED** [review]
otherwise LGTM

```
https://github.com/NixOS/nixpkgs/pull/122670

25 packages built:
ansible ansible-lint ansible_2_9 kargo python38Packages.ansible-collections python38Packages.ansible-kernel python38Packages.ansible-runner python38Packages.jxmlease python38Packages.pynetbox python38Packages.pytest-ansible python38Packages.pyxb ttp python38Packages.yangson python39Packages.ansible python39Packages.ansible-base python39Packages.ansible-collections python39Packages.ansible-kernel python39Packages.ansible-lint python39Packages.ansible-runner python39Packages.jxmlease python39Packages.pynetbox python39Packages.pytest-ansible python39Packages.pyxb python39Packages.ttp python39Packages.yangson
```

File: pkgs/development/python-modules/ansible/collections.nix (hunk 8 above)
Comment (mweinelt) [reply pkgs/development/python-modules/ansible/collections.nix:64]: There shouldn't be anything to strip in there and `find [...] -exec strip [...]` on close to 50k files adds roughly 6 minutes to the build process.

File: pkgs/development/python-modules/ansible/collections.nix [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    # ansible using overrides to enable the collections you need.
+  ] ++ lib.optionals (withJunos) [
+    # ansible_collections/junipernetworks/junos/requirements.txt
+    jxmlease
+    ncclient
+    paramiko
+    scp
+    xmltodict
+  ] ++ lib.optionals (withNetbox) [
+    # ansible_collections/netbox/netbox/pyproject.toml
```
Comment (mweinelt) [inline pkgs/development/python-modules/ansible/collections.nix:61]: Added this missing comment, nothing else.

File: pkgs/development/python-modules/ansible/collections.nix (hunk 8 above)
Comment (jonringer) [reply pkgs/development/python-modules/ansible/collections.nix:65]: oh, do you mind adding that as a comment? otherwise someone might try removing it

Comment (mweinelt) [reply pkgs/development/python-modules/ansible/collections.nix:65]: Copying from home-assistant :smile: 

```
  # don't try and fail to strip 6600+ python files, it takes minutes!
  dontStrip = true;
```

File: pkgs/development/python-modules/ansible/collections.nix [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+  # don't try and fail to strip 48000+ non strippable files, it takes >5 minutes!
+  dontStrip = true;
+
+  # difficult to test
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Radically simple IT automation";
+    homepage = "Radically simple IT automation";
```
Comment (sengaya) [inline pkgs/development/python-modules/ansible/collections.nix:73]: Should be `http://www.ansible.com`

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 122670 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>25 packages built:</summary>  <ul>    <li>ansible (ansible_2_10 ,python38Packages.ansible-base)</li>    <li>ansible-lint (python38Packages.ansible-lint)</li>    <li>ansible_2_9 (python38Packages.ansible)</li>    <li>kargo</li>    <li>python38Packages.ansible-collections</li>    <li>python38Packages.ansible-kernel</li>    <li>python38Packages.ansible-runner</li>    <li>python38Packages.jxmlease</li>    <li>python38Packages.pynetbox</li>    <li>python38Packages.pytest-ansible</li>    <li>python38Packages.pyxb</li>    <li>ttp (python38Packages.ttp)</li>    <li>python38Packages.yangson</li>    <li>python39Packages.ansible</li>    <li>python39Packages.ansible-base</li>    <li>python39Packages.ansible-collections</li>    <li>python39Packages.ansible-kernel</li>    <li>python39Packages.ansible-lint</li>    <li>python39Packages.ansible-runner</li>    <li>python39Packages.jxmlease</li>    <li>python39Packages.pynetbox</li>    <li>python39Packages.pytest-ansible</li>    <li>python39Packages.pyxb</li>    <li>python39Packages.ttp</li>    <li>python39Packages.yangson</li>  </ul></details>

File: pkgs/development/python-modules/ansible/collections.nix (hunk 10 above)
Comment (mweinelt) [reply pkgs/development/python-modules/ansible/collections.nix:73]: Copy/paste is hard. I hope we as a people will conquer this technique some day.
